include/uk/plat/paging.h

```diff
@@ -0,0 +1,279 @@
+
+/**
+ * Initializes a new page table as clone of another page table. The new page
+ * table shares the physical address range available for new allocations and
+ * mappings with the source page table.
+ *
+ * @param pt pointer to an uninitialized page table that will receive a
+ *   clone of the source page table.
+ * @param pt_src pointer to the source page table that will be cloned.
+ * @param flags clone flags (PAGE_FLAG_CLONE_*). If PAGE_FLAG_CLONE_NEW is
+ *   specified, a new (empty) top-level page table is created. Note that
+ *   this page table will be completely empty and thus do not map any code or
+ *   data segments of the kernel.
+ *
+ * @return 0 on success, a non-zero value otherwise.
+ */
+int ukplat_pt_clone(struct uk_pagetable *pt, struct uk_pagetable *pt_src,
+		    unsigned long flags);
+
+/**
+ * Frees the given page table by recursively releasing the page table hierarchy
+ * and all mapped physical memory. Mapped physical memory not belonging to the
+ * page table is not freed.
+ *
+ * @param pt the page table hierarchy to release.
+ * @param flags page flags (PAGE_FLAG_* flags). If PAGE_FLAG_KEEP_FRAMES is
+ *   specified, mapped physical memory is not freed.
+ */
+int ukplat_pt_free(struct uk_pagetable *pt, unsigned long flags);
+
+/**
+ * Performs a page table walk
+ *
+ * @param pt the page table instance on which to operate.
+ * @param vaddr the virtual address to translate.
+ * @param [in,out] level the level in which the walk should stop. Use PAGE_LEVEL
+ *   to perform a complete walk. The parameter returns the level in the page
+ *   table where the translation ended. This value might be higher than the
+ *   specified level, depending on the page table. Can be NULL in which case
+ *   PAGE_LEVEL is assumed.
+ * @param [out] pt_vaddr the virtual address of the page table where the
+ *   translation ended. Can be NULL. Can be used with ukarch_pte_write() and
+ *   PT_Lx_IDX() to update the PTE.
+ * @param [out] pte the PTE where the translation ended. Can be NULL.
+ *
+ * @return 0 on success, a non-zero value otherwise.
+ */
+int ukplat_pt_walk(struct uk_pagetable *pt, __vaddr_t vaddr,
+		   unsigned int *level, __vaddr_t *pt_vaddr, __pte_t *pte);
+
+/**
+ * Creates a mapping from a range of continuous virtual addresses to a range of
+ * physical addresses using the specified protection.
+ *
+ * @param pt the page table instance on which to operate.
+ * @param vaddr the virtual address of the first page in the new mapping.
+ * @param paddr the physical address of the memory which the virtual region is
+ *   mapped to. This parameter can be __PADDR_ANY to dynamically allocate
+ *   physical memory as needed.
+ * @param pages the number of pages in requested page size to map.
+ * @param prot the protections to set for the new mapping (PAGE_PROT_* flags).
+ * @param flags page flags (PAGE_FLAG_* flags). The page size can be specified
+ *   with PAGE_FLAGS_SIZE(). If PAGE_FLAG_FORCE_SIZE is not specified, the
```

Comment:
Small typo here, s/PAGE_FLAGS_SIZE/PAGE_FLAG_SIZE/ or update the definitions.

Also, how does the caller know know what level value to pass to the `PAGE_FLAGS_SIZE()` macro? This is architecture-specific, and at least in Arm, the page granularity is configurable. You can have: 
* 4KiB Granule: 4KiB, 2MiB, 1GiB
* 16KiB Granule: 16KiB, 32MiB, 
* 64KiB Granule: 64KiB, 512MiB

On the other hand, I don't expect we will need to support anything other than the 4KiB granule (famous last words).

Anyway, if this API was not public as this function would be called by low-level code, where we can assume it can determine somehow the available sizes. Alternatively, we can introduce a function that returns the currently supported page sizes per level. Or instead of level have the user pass definitions like #PAGE_SIZE_4KB. Sometimes you forget which level corresponds to which size, especially given that the indexing convention between AArch64 and x86 for levels is different (L0->L1->L2->L3 vs L3->L2->L1-L0).

Last but not least, should this header be part of the public API? I would expect that include/uk/ would contain higher level functions that use this API, like `mmap()` and friends.

Reply:
> Small typo here, s/PAGE_FLAGS_SIZE/PAGE_FLAG_SIZE/ or update the definitions.

I'll fix the typo 👍🏻 

> Also, how does the caller know know what level value to pass to the `PAGE_FLAGS_SIZE()` macro? This is architecture-specific, and at least in Arm, the page granularity is configurable. You can have:
> 
> * 4KiB Granule: 4KiB, 2MiB, 1GiB
> * 16KiB Granule: 16KiB, 32MiB,
> * 64KiB Granule: 64KiB, 512MiB
> 
> On the other hand, I don't expect we will need to support anything other than the 4KiB granule (famous last words).
>
> Anyway, if this API was not public as this function would be called by low-level code, where we can assume it can determine somehow the available sizes. Alternatively, we can introduce a function that returns the currently supported page sizes per level. Or instead of level have the user pass definitions like #PAGE_SIZE_4KB. Sometimes you forget which level corresponds to which size, especially given that the indexing convention between AArch64 and x86 for levels is different (L0->L1->L2->L3 vs L3->L2->L1-L0).

Yes, this is somewhat suboptimal and in fact I decided for the more abstract usage of levels here because I am aware of this configurable page size on ARM. So I did not want to globally define something like `PAGE_SIZE_4KB`. Instead, on x86 I have defined `PAGE_LEVEL`, `PAGE_LARGE_LEVEL`, and `PAGE_HUGE_LEVEL`. I thought that if you need to go for a certain page size you probably have a very good reason for this and that this will also probably happen in architecture-dependent code. So you would be able to use these more human readable `*_LEVEL` macros in that case. Otherwise, you can just leave it to the paging code to choose the largest possible size on the architecture. The only exception is that `PAGE_LEVEL` should always relate to the smallest possible page size. But of course, we could also use the terms `LARGE`, `HUGE` and perhaps at some point in the future `GIANT` in a more architecture-independent way and define these also on ARM (with different meanings for different granule configurations).

Nevertheless, you might still need all these level-based macros like `PAGE_Lx_SIZE` etc. also in generic code. But my assumption is that this will be for the most part code that uses the page table walking function (`ukplat_pt_walk`), where you get the level from the API. So you do not need to know in advance what level to select and you can use one of the `PAGE_Lx_*` macros to find out, for example, what size the page has that is mapped at the walked address.

If you would want to determine available page sizes and levels in a generic way, you would do something like this:
```C
for (int lvl = PAGE_LEVEL; lvl < PT_LEVELS; lvl++) {
    if (PAGE_Lx_HAS(lvl))
       printf("System supports page level %d with page size %d\n", lvl, PAGE_Lx_SIZE(lvl));
}
```
 
> Last but not least, should this header be part of the public API? I would expect that include/uk/ would contain higher level functions that use this API, like `mmap()` and friends.

Yes, this is intended to be UKs public paging API. Everything you build on top of it is your own business. So POSIX `mmap` will be part of the `posix-mmap` library that implements all of the POSIX memory mapping related syscalls (e.g., also `mprotect`). However, I would like to have a generic UK library that manages the virtual address space and which has an overview of existing virtual memory areas. And the `posix-mmap` library would use this library underneath. But I see `mmap`-type functionality as an optional feature because as we discussed in the meeting you might want to go with a more fixed layout, in which case you only have this paging API or not even that if you go with a fully static page table.

Reply:
> Yes, this is somewhat suboptimal and in fact I decided for the more abstract usage of levels here because I am aware of this configurable page size on ARM. So I did not want to globally define something like PAGE_SIZE_4KB. Instead, on x86 I have defined PAGE_LEVEL, PAGE_LARGE_LEVEL, and PAGE_HUGE_LEVEL. I thought that if you need to go for a certain page size you probably have a very good reason for this and that this will also probably happen in architecture-dependent code. So you would be able to use these more human readable *_LEVEL macros in that case. Otherwise, you can just leave it to the paging code to choose the largest possible size on the architecture. The only exception is that PAGE_LEVEL should always relate to the smallest possible page size. But of course, we could also use the terms LARGE, HUGE and perhaps at some point in the future GIANT in a more architecture-independent way and define these also on ARM (with different meanings for different granule configurations).
>
> Nevertheless, you might still need all these level-based macros like PAGE_Lx_SIZE etc. also in generic code. But my assumption is that this will be for the most part code that uses the page table walking function (ukplat_pt_walk), where you get the level from the API. So you do not need to know in advance what level to select and you can use one of the PAGE_Lx_* macros to find out, for example, what size the page has that is mapped at the walked address.

I agree for the most part, although something like GIANT would still break if higher granules only provide two possible sizes. In any case, as granularity is controlled by the platform, and may be chosen either based on a KConfig option on compile time, or based on processor capabilities on runtime, one would need to write some boilerplate if they need to support different granularities, whether we express this information on levels or numbers, so it's okay. If different granularities are ever required to be supported in Unikraft, that is. So let's keep it as it is then :)

> Yes, this is intended to be UKs public paging API. Everything you build on top of it is your own business. So POSIX mmap will be part of the posix-mmap library that implements all of the POSIX memory mapping related syscalls (e.g., also mprotect). However, I would like to have a generic UK library that manages the virtual address space and which has an overview of existing virtual memory areas. And the posix-mmap library would use this library underneath. But I see mmap-type functionality as an optional feature because as we discussed in the meeting you might want to go with a more fixed layout, in which case you only have this paging API or not even that if you go with a fully static page table.

Okay, fair enough 👍🏼